pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,259 @@
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+
+import "../BaseWeightedPool.sol";
+import "./WeightCompression.sol";
+
+/**
+ * @dev Weighted Pool with mutable weights, designed to support investment use cases: large token counts,
+ * rebalancing through gradual weight updates, and enabling/disabling trading.
+ */
+contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
+    // The Pause Window and Buffer Period are timestamp-based: they should not be relied upon for sub-minute accuracy.
+    // solhint-disable not-rely-on-time
+
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+    using WeightCompression for uint256;
+
+    // State variables
+
+    // Cached here to avoid calling getTokens on the pool, which would be very gas-intensive for large numbers of tokens
+    // Can only change if tokens are added/removed
+    uint256 private _tokenCountCache;
+
+    // Store scaling factor and start/end weights for each token
+    // Mapping should be more efficient than trying to compress it further
+    // into a fixed array of bytes32 or something like that, especially
+    // since tokens can be added/removed - and re-ordered in the process
+    // For each token, we store:
+    // [ 155 bits|   5 bits |  32 bits   |   64 bits    |
+    // [ unused  | decimals | end weight | start weight |
+    // |MSB                                          LSB|
+    mapping(IERC20 => bytes32) private _poolState;
+
+    uint256 private constant _START_WEIGHT_OFFSET = 0;
+    uint256 private constant _END_WEIGHT_OFFSET = 64;
+    uint256 private constant _DECIMAL_DIFF_OFFSET = 96;
+
+    // Time travel comment
+    // [ 192 bits | 32 bits  |  32 bits   |
+    // [  unused  | end time | start time |
+    // |MSB                            LSB|
+    //bytes32 private _gradualUpdateTimestamps;
+
+    //uint256 private constant _START_TIME_OFFSET = 0;
+    //uint256 private constant _END_TIME_OFFSET = 32;
+
+    constructor(
+        IVault vault,
+        string memory name,
+        string memory symbol,
+        IERC20[] memory tokens,
+        uint256[] memory normalizedWeights,
+        address[] memory assetManagers,
+        uint256 swapFeePercentage,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration,
+        address owner
+    )
+        BaseWeightedPool(
+            vault,
+            name,
+            symbol,
+            tokens,
+            assetManagers,
+            swapFeePercentage,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+    {
+        uint256 numTokens = tokens.length;
+        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length, assetManagers.length);
+
+        _tokenCountCache = numTokens;
+
+        // I'm time-traveling a bit here - storing the weights in a form where they can be changed
+        uint256 currentTime = block.timestamp;
```

Comment:
Not sure I follow this comment, what do you mean by time-traveling? This seems consistent with what we do in LBPs: an instant weight change (start and end time are the same).

Reply:
I just mean I'm putting code in that anticipates the next PR